[ckey] re-introduce context key optimization

vickenty opened the pull request:
### What does this PR do?

Bring back context key optimization from #8529.

Re-arrange tags during context key generation, so the resulting set is unique and can be used for context entry.

Add fallback for metrics with more than 512 tags.

More tests, check that tags are de-duplicated by each of three algorithms.

Properly handle hash collisions and zero hash value.

### Motivation

Less CPU and memory load.

### Describe how to test your changes

```
log_level: debug
log_payloads: true
dogstatsd_socket: /tmp/dsd.sock
```

Check that metrics with 1, 10, 50, 100, 600 tags are sent to the back-end.

```
python -c 'print("my.stuff:1|c|#"+",".join(["t%d" % x for x in range(600)]))' | nc -uU /tmp/dsd.sock
```

Send a new metric with duplicate tags (new metric to avoid re-using context); check that metric is sent to the backend with unique tags. Send the same metric with unique tags, should be the same metric sent to the back end.

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

vickenty committed:
Revert "Revert "[aggregator/ckey] implement a faster hash to reduce CPU and RAM usage (#8529)" (#8678)"

This reverts commit 5845856cb3eeb64a54402e0341c792df4fa7f62d.

vickenty committed:
[ckey] make Generate accept TagsBuilder

TagsBuilder is mutable slice wrapper, and this will allow us to
de-duplicate tags in it at the same as generating the context key.

vickenty committed:
[ckey] de-duplicate tags during ckey generation

Shift tags around in the tags builder and truncate it at the end.

vickenty committed:
[ckey] add fallback for len(tags) > 512

vickenty committed:
[ckey] check for hash collisions

Skip tags if their hashes and string values are the same, instead of
just checking hashes.

For bruteforce method, seen and tags arrays are filled at the same time,
so adding a check is straightforward.

For hashset method, a new array seenIdx is added, and we store the
index of the tag that corresponds to the current hashset slot.

This also handles tags which hash to zero correctly.

djmitche commented on pkg/aggregator/ckey/key.go:
```diff
@@ -27,49 +24,131 @@ import (
+	// seenIdx is the index of the tag stored in the hashset
+	seenIdx []int
+	// empty is an empty hashset with all values set to `blank`, to reset `seenIdx`
+	empty []int
```
I asked last time this was changed, but .. I think go uses `memset` properly to clear arrays, which should be substantially faster than copying a known-zero memory region over.

djmitche replied:
Ah, here's the link -- https://github.com/golang/go/commit/f03c9202c43e0abb130669852082117ca50aa9b1

djmitche replied:
We talked about this and it turns out that because this is in a struct, the compiler can't prove that the array is not aliased, so it can't use memclr on it.

djmitche commented on pkg/aggregator/ckey/key.go:
```diff
@@ -27,49 +24,131 @@ import (
+	// 16 and less than 512 tags.
+	seen []uint64
+	// seenIdx is the index of the tag stored in the hashset
+	seenIdx []int
```
I think you could save an allocation by using arrays here, instead of slices?

djmitche commented on pkg/aggregator/ckey/key.go:
```diff
@@ -27,49 +24,131 @@ import (
+		copy(g.seenIdx, g.empty)
+		for i := range tags {
+			h := murmur3.StringSum64(tags[i])
+			j := h & (hashSetSize - 1) // address this hash into the hashset
```
Worth noting that `hashSetSize` must be a power of two?

djmitche commented on pkg/aggregator/ckey/key_test.go:
```diff
@@ -49,21 +51,100 @@ func TestCompare(t *testing.T) {
 }
 
-func genTags(count int) []string {
+func TestTagsOrderAndDupsDontMatter(t *testing.T) {
```
In [this experiment](https://github.com/djmitche/tagset-experiment/blob/e4085f2250a4357338e5d8cb0f49375d0f30f1d9/tagset/tagset_test.go#L92-L138) I did some seeded randomized trials as a kind of lightweight fuzzing.  Maybe worth adding a similar thing here?

vickenty committed:
[ckey] use arrays to reduce bounds checks

Using arrays instead of dynamic slices, lets go drop some of the bounds
checks.

vickenty committed:
[ckey] use smaller index array

We only store indexes up to 512, and using a smaller type means less
bytes need to be moved to initialize `seenIdx`.

vickenty committed:
[ckey] size hashset according to tags

Initialize a part of the hashset, big enough to process incoming tags,
but not larger. Has most of positive effect when number of tags <<
hashSetSize, which makes hashset method faster than bruteforce earler.

vickenty committed:
[ckey] add comment about hash set size

vickenty committed:
go fmt

djmitche reviewed (approved):
Awesome :D
